internal/app/machined/pkg/controllers/cri/image_cache_config.go

```diff
@@ -287,16 +293,116 @@ func (ctrl *ImageCacheConfigController) createVolumeConfigDisk(ctx context.Conte
 	})
 }
 
-func (ctrl *ImageCacheConfigController) getImageCacheRoot(ctx context.Context, r controller.Reader, volumeID string) (optional.Optional[string], bool, error) {
-	volumeStatus, err := safe.ReaderGetByID[*block.VolumeStatus](ctx, r, volumeID)
-	if err != nil {
-		if state.IsNotFoundError(err) {
-			return optional.None[string](), false, nil
+type imageCacheVolumeStatus struct {
+	roots      []string
+	allReady   bool
+	copyStatus cri.ImageCacheCopyStatus
+}
+
+//nolint:gocyclo,cyclop
+func (ctrl *ImageCacheConfigController) analyzeImageCacheVolumes(ctx context.Context, logger *zap.Logger, r controller.Reader) (*imageCacheVolumeStatus, error) {
+	volumeIDs := []string{VolumeImageCacheDISK, VolumeImageCacheISO} // prefer disk cache over ISO cache
+	volumeStatuses := make([]*block.VolumeStatus, 0, len(volumeIDs))
+
+	for _, volumeID := range volumeIDs {
+		volumeStatus, err := safe.ReaderGetByID[*block.VolumeStatus](ctx, r, volumeID)
+		if err != nil {
+			if state.IsNotFoundError(err) {
+				// wait for volume statuses to be present
+				return &imageCacheVolumeStatus{}, nil
+			}
+
+			return nil, fmt.Errorf("error getting volume status: %w", err)
+		}
+
+		volumeStatuses = append(volumeStatuses, volumeStatus)
+	}
+
+	// we need to ensure that we first wait for the ISO to be either missing or ready,
+	// so that we can make a decision on copying the image cache from an ISO to the disk volume
+	var isoStatus, diskStatus block.VolumePhase
+
+	for _, volumeStatus := range volumeStatuses {
+		switch volumeStatus.Metadata().ID() {
+		case VolumeImageCacheISO:
+			isoStatus = volumeStatus.TypedSpec().Phase
+		case VolumeImageCacheDISK:
+			diskStatus = volumeStatus.TypedSpec().Phase
+		}
+	}
+
+	if isoStatus != block.VolumePhaseMissing && isoStatus != block.VolumePhaseReady {
+		return &imageCacheVolumeStatus{}, nil
+	}
+
+	isoPresent := isoStatus == block.VolumePhaseReady
+	diskMissing := diskStatus == block.VolumePhaseMissing
+
+	roots := make([]string, 0, len(volumeIDs))
+
+	var (
+		allReady, isoReady, diskReady bool
+		copySource, copyTarget        string
+	)
+
+	// analyze volume statuses, and build the roots
+	for _, volumeStatus := range volumeStatuses {
+		// mount as rw only disk cache if the ISO cache is present
+		root, ready, err := ctrl.getImageCacheRoot(ctx, r, volumeStatus, !(volumeStatus.Metadata().ID() == VolumeImageCacheDISK && isoPresent))
+		if err != nil {
+			return nil, fmt.Errorf("error getting image cache root: %w", err)
+		}
+
+		if ready {
+			switch volumeStatus.Metadata().ID() {
+			case VolumeImageCacheISO:
+				isoReady = true
+				copySource = root.ValueOr("")
+			case VolumeImageCacheDISK:
+				diskReady = true
+				copyTarget = root.ValueOr("")
+			}
+		}
+
+		allReady = allReady && ready
```

Comment:
Since there could only be `VolumeImageCacheDISK` and `VolumeImageCacheISO` volume statuses, doesn't it make sense to move and redo `allReady = allReady && ready`  as `allReady = isoReady && diskReady`? Or I'm missing something.

Reply:
I guess we can also define `copyTarget` and `copySource` as `optional.Optional[string]` and drop `diskReady` and `isoReady` entirely.

Reply:
I want to leave some room for expansion and support other read-only sources